daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala

```diff
@@ -235,11 +339,291 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
+             in let cId2: ContractId M:T = COERCE_CONTRACT_ID @M:Dummy @M:T cId1
+             in fetch @M:T cId2""",
+            Array(SParty(alice)),
+            alice,
+          )
+          inside(res) {
+            case Success(Left(SErrorDamlException(IE.WronglyTypedContract(_, T, Dummy)))) =>
+          }
+          msgs shouldBe Seq.empty
+        }
+      }
+
+      // TEST_EVIDENCE: Semantics: Evaluation order of fetch of an unknown contract
+      "unknown contract" in {
+        val (res, msgs) = evalUpdateApp(
+          pkgs,
+          e"""\(cId: ContractId M:T) -> fetch @M:T cId""",
+          Array(SContractId(cId)),
+          alice,
+          getContract = PartialFunction.empty,
+        )
+        inside(res) { case Failure(SpeedyTestLib.UnknownContract(`cId`)) => }
+        msgs shouldBe Seq("getContract")
+      }
+
+    }
+
+    "fetch_by_key" - {
 
+      "a global contract" - {
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of successful fetch_by_key of a global contract
+        "success" in {
+          val (res, msgs) = evalUpdateApp(
+            pkgs,
+            e"""\(sig: Party) -> fetch_by_key @M:T (M:toKey sig)""",
+            Array(SParty(alice)),
+            alice,
+            getContract = getContract,
+            getKey = getKey,
+          )
+          inside(res) { case Success(Right(_)) => }
+          msgs shouldBe globalFetchByKeyTraces
+        }
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of fetch_by_key of an inactive global contract
+        "inactive contract" in {
+
+          val (res, msgs) = evalUpdateApp(
+            pkgs,
+            e"""\(cId: ContractId M:T) (sig: Party) -> 
+             ubind x: Unit <- exercise @M:T Archive cId sig
+             in fetch_by_key @M:T (M:toKey sig)""",
+            Array(SContractId(cId), SParty(alice)),
+            alice,
+            getContract = getContract,
+            getKey = getKey,
+          )
+          inside(res) { case Success(Left(SErrorDamlException(IE.ContractKeyNotFound(key)))) =>
+            key.templateId shouldBe T
+          }
+          msgs shouldBe globalFetchByIdTraces :++ Set("archive", "maintainers", "getKey")
+        }
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of fetch of a global contract with authorization failure
+        "authorization failures" in {
+          val (res, msgs) = evalUpdateApp(
+            pkgs,
+            e"""\(sig: Party) -> fetch_by_key @M:T (M:toKey sig)""",
+            Array(SParty(alice)),
+            charlie,
+            getContract = getContract,
+            getKey = getKey,
+          )
+          inside(res) { case Success(Left(SErrorDamlException(IE.FailedAuthorization(_, _)))) => }
+          msgs shouldBe globalFetchByKeyTraces
+        }
+      }
+
+      "a local contract" - {
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of successful fetch_by_key of a local contract
```

Comment:
similar comment as above, it doesn’t matter if it’s local or not it matters that it’s cached. The second fetch by key of a global contract will do the same and we should test that.

Reply:
I will take add test for cached/non-cached contract in separate PR.